akka-actor/src/main/scala/akka/dispatch/affinity/AffinityPool.scala

```diff
@@ -0,0 +1,341 @@
+
+  if (parallelism <= 0)
+    throw new IllegalArgumentException("Size of pool cannot be less or equal to 0")
+
+  // Held while starting/shutting down workers/pool in order to make
+  // the operations linear and enforce atomicity. An example of that would be
+  // adding a worker. We want the creation of the worker, addition
+  // to the set and starting to worker to be an atomic action. Using
+  // a concurrent set would not give us that
+  private val bookKeepingLock = new ReentrantLock()
+
+  // condition used for awaiting termination
+  private val terminationCondition = bookKeepingLock.newCondition()
+
+  // indicates the current state of the pool
+  @volatile final private var poolState: PoolState = Running
+
+  private final val workQueues = Array.fill(parallelism)(new BoundedTaskQueue(affinityGroupSize))
+  private final val workers = mutable.Set[ThreadPoolWorker]()
+
+  // a counter that gets incremented every time a task is queued
+  private val executionCounter: AtomicInteger = new AtomicInteger(0)
+  // maps a runnable to an index of a worker queue
+  private val runnableToWorkerQueueIndex: java.util.Map[Int, Int] = new ConcurrentHashMap[Int, Int]()
+
+  /**
+   *
+   * Given a Runnable, it returns the queue that should
+   * be used for scheduling it. This way we ensure that
+   * the each Runnable is associated with the the same
+   * queue and thereby the same [[ThreadPoolWorker]].
+   * For the purpose of evenly distributing [[Runnable]]
+   * instances across the workers we use executionCounter
+   * to produce a round robbin - like scheme that cycles
+   * from 0 to n-1 where n is the number of queues that
+   * this executor maintains. When we compute the index
+   * of the queue for a particular [[Runnable]] instance,
+   * the hashCode of the [[Runnable]] is mapped to the
+   * queue index. From then on each time this Runnable
+   * is scheduled, it shall go to the same queue.
+   *
+   * A race condition might occur due to the absence of
+   * atomicity when invoking incrementAndGet
+   * on the [[executionCounter]] and inserting the result
+   * into the [[runnableToWorkerQueueIndex]]. This is ok.
+   * Worst case scenario is that there will be AT MOST 2
+   * distinct executions of the same Runnable instance that
+   * will map to different queues. This wil be negligible
+   * for performance and is a compromise that can be allowed.
+   * The alternative is holding a lock, which is not
+   * an attractive proposition.
+   *
+   * @param command the [[Runnable]] that needs to be scheduled
+   * @return the [[BoundedTaskQueue]] to which this runnable should go
+   */
+  private def getQueueForRunnable(command: Runnable) = {
+    def getNext = executionCounter.incrementAndGet() & (parallelism - 1)
+    workQueues(runnableToWorkerQueueIndex.getOrElseUpdate(command.hashCode, getNext))
```

Comment:
@patriknw I certainly share your skepticism regarding this. However, maybe take a look at [this discussion](https://github.com/akka/akka/commit/111e04f580ec4d693aa3f148abeaf9c2ad6b5c2c#commitcomment-22648383) and the associated benchmarks that I have provided. Seems for small number of actors it is worth paying the cost of the map (allocation, contention, etc.) in order to ensure even work distribution. I am quite keen on my last proposition where we use a hybrid strategy up until a certain number of distinct actors (say 2048). What do you think ?